README.md
# forecasts

## Installation

Download from https://github.com/ougfh/forecasts.

## How to run

Build an uberjar:

    $ clojure -A:uberjar

Run that uberjar with darksky API key:

    $ DARKSKY_API_KEY=api_key java -jar forecasts.jar

with passed darksky API key as enviroment variable.

## Options

Options that can be passed:

- `--city` city name
- `--lat` latitude, default 60.59329987
- `--lng` longitude, default -1.44250533

## Test

Run the project's tests:

    $ clojure -A:test:runner

## Examples

To see weaher reports from Moscow by city name, run:

    $ DARKSKY_API_KEY=api_key java -jar forecasts.jar --city Moscow

or by coordinates:

    $ DARKSKY_API_KEY=api_key java -jar forecasts.jar --lat 55.751244 --lng 37.618423

## TODO

- Find the hottest month of 2019
- Use `exclude` parameter in the request to darksy
- Parameters validation
- Improve output, use polymorphism instead of println to dynamically adjust the output
- Separate strings creation and data extraction; can be used translation library to produce an output string
- Java startup time, too slow for CLI
- Use optional parameters in darksky API, e.x. language, units
- Use HTTP compression
- Check for the presence of data in response before use

## License

Copyright © 2020 Rodionovaanastasia

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-08-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-08-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/forecasts/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/forecasts/compare/0.1.0...0.1.1


test/forecasts/core_test.clj
(ns forecasts.core-test
  (:require [forecasts.core :as sut]
            [clojure.test :refer [are is testing deftest]]
            [clojure.string :as cstr]))

(def forecast-example
  {:currently {:summary            "Drizzle"
               :precip-probability 0.9}
   :hourly    {:summary "Rain."}
   :daily     {:data [{:icon "rain"}
                      {:icon "rain"}
                      {:icon "partly-cloudy-day"}]}})

(deftest summary-report
  (let [report (sut/get-summary-report forecast-example)]
    (testing "target keys are in report"
      (are [x y] (cstr/includes? x y)
        report (get-in forecast-example [:currently :summary])
        report (str (get-in forecast-example [:currently :precip-probability]))
        report (get-in forecast-example [:hourly :summary])))))

(deftest daily-icons-report
  (let [report             (sut/get-daily-icons-report forecast-example)
        expected-icons-str "2 days rain, 1 days partly-cloudy-day"]
    (is (cstr/includes? report expected-icons-str))))


src/forecasts/core.clj
(ns forecasts.core
  (:require [hato.client :as client]
            [clojure.data.json :as json]
            [camel-snake-kebab.core :as csk]
            [clojure.string :as cstr]
            [java-time :as time]
            [clojure.tools.cli :as cli])
  (:gen-class))

(def darksky-api-key (System/getenv "DARKSKY_API_KEY"))
(def forecasts-url (partial format "https://api.darksky.net/forecast/%s/%s,%s"))
(def time-machine-request-url (partial format "https://api.darksky.net/forecast/%s/%s,%s,%s"))
(def cities-url "https://raw.githubusercontent.com/lutangar/cities.json/master/cities.json")

(def summary-template (partial format "Current weather - %s, Today we will see - %s with a %s%% chance of rain."))
(def hottest-day-template (partial format "This week the hottest day will be %s."))
(def icon-frequencies-template (partial format "This week we should have %s."))

(def temperature-max-key :temperature-max)

(defn get-forecast
  "returns the current weather forecast for the next week."
  [{:keys [lat lng]}]
  (when darksky-api-key
    (-> (forecasts-url darksky-api-key lat lng)
        client/get
        :body
        (json/read-str :key-fn csk/->kebab-case-keyword))))

(defn get-time-machine-forecast
  "returns the observed or forecast weather conditions for a date in the past or future."
  [{:keys [lat lng time]}]
  (-> (client/get (time-machine-request-url darksky-api-key lat lng time))
      :body
      (json/read-str :key-fn csk/->kebab-case-keyword)))

(defn get-cities-list
  "returns list of {:contry :name :lat :lng }"
  []
  (-> (client/get cities-url)
      :body
      (json/read-str :key-fn csk/->kebab-case-keyword)))

(defn equal-strings? [a b]
  (= (cstr/lower-case a) (cstr/lower-case b)))

(defn find-city-coordinates [city-name]
  (->> (get-cities-list)
       (filter #(equal-strings? city-name (:name %)))
       first))

(defn find-value-with-max-key [coll key]
  (last (sort-by key coll)))

(defn get-day-of-week [epoch-in-sec zone]
  (time/day-of-week
   (time/as
    (time/offset-date-time (time/instant (* epoch-in-sec 1000)) zone)
    :day-of-week)))

(defn extract-summary-fields [resp]
  {:currently-summary            (get-in resp [:currently :summary])
   :hourly-summary               (get-in resp [:hourly :summary])
   :currently-precip-probability (get-in resp [:currently :precip-probability])})

(defn get-summary-report [forecast]
  (let [{:keys [currently-summary hourly-summary currently-precip-probability]}
        (extract-summary-fields forecast)]
    (summary-template currently-summary
                      hourly-summary
                      currently-precip-probability)))

(defn get-hottest-day-report [forecast]
  (let [daily-forecast (get-in forecast [:daily :data])
        hottest-day    (find-value-with-max-key daily-forecast temperature-max-key)
        day-of-week    (get-day-of-week (:time hottest-day) (:timezone forecast))]
    (hottest-day-template (cstr/capitalize day-of-week))))

(defn get-daily-icons-report [forecast]
  (let [daily-forecast (get-in forecast [:daily :data])
        icons          (frequencies (map :icon daily-forecast))
        icons-str      (map #(str (second %) " days " (first %)) icons)]
    (icon-frequencies-template (cstr/join ", " icons-str))))

(defn do-reports [coordinates]
  (let [forecast    (get-forecast coordinates)
        summary     (get-summary-report forecast)
        hottest-day (get-hottest-day-report forecast)
        daily-icons (get-daily-icons-report forecast)]
    (println summary)
    (println hottest-day)
    (println daily-icons)))

(def cli-options
  [[nil "--city CITY" "City name"
    :parse-fn identity]
   [nil "--lat LAT" "Latitude"
    :default 60.59329987
    :parse-fn read-string]
   [nil "--lng LNG" "Longitude"
    :default -1.44250533
    :parse-fn read-string]
   ["-h" "--help"]])

(defn -main
  [& args]
  (let [{:keys [city lat lng]}
        (:options (cli/parse-opts args cli-options))

        coordinates (if city
                      (find-city-coordinates city)
                      {:lat lat :lng lng})]
    (if (and city (nil? coordinates))
      (println "No results found for " city)
      (do-reports coordinates))))
